### PR TITLE
Issue 366 - Bouncer worker Improvement

### LINKED ISSUE #366: Refactor bouncer_worker
Refactoring effort in bouncerworker

- [x] Refresh it with ES6 syntax
- [x] Restructure functions to be more modular
- [x] modularise toy project function so it can be called separately easier (so we can trigger rerun where necessary)
- [x] New Unity queue:
    - [x] code to support new unity queue
    - [x] Add a pending unity build flag when the bouncer has finished but unity hasn't started
    - [x] Update bouncer wrapper to look for revisions that has this flag
- [x] refactor the way to check if we have all the info to process the specific queued tasks
- [x] A way to run the script that will just exit after processing 1 single task
- [x] update package.json with new commands
- [x] Add linter
- [x] send placeholder for file share directory instead of hard coding `/sharedData`
- [x] Add bouncer version in the bouncer log
- [x] Clean up logging
    - [x] option to disable logging to file
    - [x] add label and better structure of the logging 
    - [x] https://github.com/3drepo/3drepobouncer/issues/325
- [x] Ensure io understands the new error code
- [x] Change travis to run with nodejs and c++ (use matrix)
- [x] update versioning script to update package.json in this folder
- [x] Add readme on bouncer worker

### PR DESCRIPTION
**Pull request by carmenfan**
This resolves #366 

This also relies on https://github.com/3drepo/3drepo.io/issues/2375 and https://github.com/3drepo/3drepounity/issues/329

See issue itself for the list of changes. See the new [Bouncer worker readme](https://github.com/3drepo/3drepobouncer/blob/046717d67fa225d0c4d40203f434b19c92433fed/tools/bouncer_worker/README.md) for new usage details

#### Deployment note
Unity is now separated into a separate queue, meaning you will need to: 
  1. add additional queue in the bouncer_worker config
  2. configure rabbitmq with this queue + give the workers the right permissions to access them (not needed if you're running `guest:guest@localhost`).

### CODE REVIEW

File: tools/bouncer_worker/src/tasks/importToy.js [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+const logger = require('../lib/logger');
+
+const logLabel = { label: 'TOY' };
+
+const accumulateCollectionFiles = (modelDir, modelId) => {
+	const importCollectionFiles = {};
+
+	fs.readdirSync(modelDir).forEach((file) => {
+		// remove '.json' in string
+		let collectionName = file.split('.');
```
Comment (Charence) [inline tools/bouncer_worker/src/tasks/importToy.js:31]: Would it not be better to do `.replace(/\.json$/,"")` instead? At the moment, popping will always remove the last element even if it isn't `json`.

**Review by Charence — CHANGES_REQUESTED** [review]

**Review by Charence — APPROVED** [review]

**Comment by keifgwinn** [comment]
I configured a jobq server by not including modelq or unityq details, but I get an error message about that

2021-01-22T16:52:28.820Z [info] [AMQP]  Waiting for messages in jobq...
2021-01-22T16:52:28.820Z [error] [MODELQ]       rabbitmq.model_queue is not specified!
2021-01-22T16:52:28.820Z [error] [UNITYQ]       rabbitmq.unity_queue is not specified!

**Comment by carmenfan** [comment]
after chat with @keifgwinn , extra things to do:
- [ ] Add code to be able to processing X number of tasks before quitting
- [ ] Replace command option `--runOnceOnQueue` with `--exitAfter x` and `--queue [job|model|unity]`

**Comment by keifgwinn** [comment]
one more nice thing, now I'm slinging a lot of versions around it would be nice if it could include the version number in the text log when booting

**Comment by carmenfan** [comment]
--version will print the version number

**Comment by keifgwinn** [comment]
❯ k logs issue-2375-unityq-zklls-6dtnv

yarn run v1.22.5
$ node src/scripts/bouncer_worker.js --queue unity --exitAfter 1
Setting umask: 2
2021-01-26T11:01:31.710Z [info] [INIT]  Checking status of client...
2021-01-26T11:01:31.711Z [info] [INIT]  Executing command: /home/bouncer/oda/teighaLinuxGCC5.3/bin/lnxX64_5.3dll/3drepobouncerClient /home/bouncer/3drepobouncer/tools/bouncer_worker/config.json test
2021-01-26T11:01:32.003Z [verbose] [INIT]       [STDOUT]: [26-01-2021 11:01:31][INFO]: Log file registered: ./log/
[26-01-2021 11:01:31][INFO]: 3D Repo Bouncer Version: 4.13.1
[26-01-2021 11:01:31][INFO]: Operation: test
[26-01-2021 11:01:31][INFO]: Successfully connected to the 10.0.1.188:27017
[26-01-2021 11:01:31][INFO]: adminUser is authenticated to 10.0.1.188:27017
[26-01-2021 11:01:31][INFO]: successfully connected to the database!
[26-01-2021 11:01:31][INFO]: Process completed, returning with error code: 0

2021-01-26T11:01:32.008Z [info] [INIT]  Command executed. Code: 0
2021-01-26T11:01:32.008Z [info] [INIT]  Bouncer call passed
2021-01-26T11:01:32.009Z [info] [AMQP]  Connecting to amqp://bouncerWorker:bouncerWorker@issue-2375-rabbitmq-headless:5672...
2021-01-26T11:01:32.094Z [info] [AMQP]  Processing 1 tasks on unityq...
2021-01-26T11:01:32.101Z [info] [AMQP]  [1/1] Received {"value":0,"database":"internal_3DRepo","project":"c8991ca0-5fbc-11eb-a876-e1140b99a65b","user":"internal_3DRepo"} from unityq
2021-01-26T11:01:32.101Z [info] [AMQP]  [1/1] Sending to reply to callbackq: {"status":"Generating Bundles","database":"internal_3DRepo","project":"c8991ca0-5fbc-11eb-a876-e1140b99a65b"}
2021-01-26T11:01:32.103Z [info] [UNITY]         Executing command: /home/bouncer/3drepounity/AssetBundleCreator/buildBundle_linux.sh /home/bouncer/3drepounity/AssetBundleCreator /home/bouncer/3drepobouncer/tools/bouncer_worker/config.json internal_3DRepo c8991ca0-5fbc-11eb-a876-e1140b99a65b /efs/queue/c8da1d40-5fbc-11eb-8e2a-75d0d068eec3/
2021-01-26T11:05:27.649Z [info] [UNITY]         Command executed. Code: 0
2021-01-26T11:05:27.649Z [info] [AMQP]  [1/1] Sending to reply to callbackq: {"value":0,"database":"internal_3DRepo","project":"c8991ca0-5fbc-11eb-a876-e1140b99a65b","user":"internal_3DRepo"}
2021-01-26T11:05:35.030Z [error] [AMQP]         Connection error: write ECONNRESET
2021-01-26T11:05:35.031Z [info] [AMQP]  Connection closed.
(node:33) UnhandledPromiseRejectionWarning: IllegalOperationError: Channel closed
    at Channel.<anonymous> (/home/bouncer/3drepobouncer/tools/bouncer_worker/node_modules/amqplib/lib/channel.js:154:11)
    at Channel.C.closeBecause (/home/bouncer/3drepobouncer/tools/bouncer_worker/node_modules/amqplib/lib/channel.js:206:8)
    at /home/bouncer/3drepobouncer/tools/bouncer_worker/node_modules/amqplib/lib/channel_model.js:78:17
    at tryCatcher (/home/bouncer/3drepobouncer/tools/bouncer_worker/node_modules/bluebird/js/release/util.js:16:23)
    at Function.Promise.fromNode.Promise.fromCallback (/home/bouncer/3drepobouncer/tools/bouncer_worker/node_modules/bluebird/js/release/promise.js:209:30)
    at Channel.C.close (/home/bouncer/3drepobouncer/tools/bouncer_worker/node_modules/amqplib/lib/channel_model.js:77:18)
    at executeTasks (/home/bouncer/3drepobouncer/tools/bouncer_worker/src/lib/queueHandler.js:108:16)
(node:33) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:33) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
❯


in this situation I want it to throw and exit, as what happened is that it is still running, and holding a space on the processing as ther'es a limit of 4 unity workers at a time, which this is counting towards

**Comment by keifgwinn** [comment]
have a similar thing at the end of modelq if it's waiting for 60 seconds and hits it

2021-01-26T11:05:35.028Z [error] [AMQP]         Connection error: write ECONNRESET
2021-01-26T11:05:35.030Z [info] [AMQP]  Connection closed.
(node:33) UnhandledPromiseRejectionWarning: IllegalOperationError: Channel closed
    at Channel.<anonymous> (/home/bouncer/3drepobouncer/tools/bouncer_worker/node_modules/amqplib/lib/channel.js:154:11)
    at Channel.C.closeBecause (/home/bouncer/3drepobouncer/tools/bouncer_worker/node_modules/amqplib/lib/channel.js:206:8)
    at /home/bouncer/3drepobouncer/tools/bouncer_worker/node_modules/amqplib/lib/channel_model.js:78:17
    at tryCatcher (/home/bouncer/3drepobouncer/tools/bouncer_worker/node_modules/bluebird/js/release/util.js:16:23)
    at Function.Promise.fromNode.Promise.fromCallback (/home/bouncer/3drepobouncer/tools/bouncer_worker/node_modules/bluebird/js/release/promise.js:209:30)
    at Channel.C.close (/home/bouncer/3drepobouncer/tools/bouncer_worker/node_modules/amqplib/lib/channel_model.js:77:18)
    at executeTasks (/home/bouncer/3drepobouncer/tools/bouncer_worker/src/lib/queueHandler.js:108:16)
(node:33) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:33) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.

**Review by keifgwinn — APPROVED** [review]
we've been fixing issues as I've been reviewing, and now toy is working, plus extensive testing on the new modes, I'm happy to merge.

File: bouncer/src/repo/lib/repo_log.cpp [hunk 2]
```diff
@@ -78,21 +78,21 @@ void RepoLog::logToFile(const std::string &filePath)
...
 		boost::log::keywords::auto_flush = true,
 		boost::log::keywords::format = (
-		boost::log::expressions::stream
-		<< "[" << boost::log::expressions::format_date_time(timestamp, "%Y-%m-%d %H:%M:%S") << "]"
-		<< "(" << threadid << ") <" << boost::log::trivial::severity
-		<< ">"
-		<< ": " << boost::log::expressions::smessage
-		));
+			boost::log::expressions::stream
+			<< "[" << boost::log::expressions::format_date_time(timestamp, "%Y-%m-%d %H:%M:%S") << "]"
```
[the remark this answers is not in the document]
Comment (keifgwinn) [reply bouncer/src/repo/lib/repo_log.cpp:91]: this is the only thing that I'm unhappy about =D

[the remark this answers is not in the document]
Comment (carmenfan) [reply bouncer/src/repo/lib/repo_log.cpp:91]: wtf that's suppose to fix it

**Pull request merged**